File: 24二叉树中和为某一值的路径.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def FindPath(self, root, expectNumber):
        if root == None:
            return []
        res = []
        self.sums = expectNumber
        self.DFS(root, res, [root.val])
        return res
    
    def DFS(self, root, res, path):
        if root.left is None and root.right is None and sum(path) == self.sums:
            res.append(path)
        if root.left is not None:
            self.DFS(root.left, res, path+[root.left.val])
        if root.right is not None:
            self.DFS(root.right, res, path+[root.right.val])

File: 剑指offer全解/24二叉树中和为某一值的路径.java
import java.util.ArrayList;
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;
    }

}
*/
// By DFS
public class Solution {
    public ArrayList<ArrayList<Integer>> FindPath(TreeNode root,int target) {
        if(root == null) return new ArrayList();
        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        findPathDFS(root, target, res, new ArrayList<>());
        return res;
    }
    public void findPathDFS(TreeNode root, int sum, ArrayList<ArrayList<Integer>> res, ArrayList<Integer> list) {
        if(root == null) return;
        
        list.add(root.val);
        sum -= root.val;
        
        if(root.left == null && root.right == null) {
            if(sum == 0) {
                res.add(new ArrayList(list));
            }
            list.remove(list.size()-1);
            return;
        }
        findPathDFS(root.left, sum, res, list);
        findPathDFS(root.right, sum, res, list);
        list.remove(list.size() - 1);
        return;
    }
}